File: 4.median-of-two-sorted-arrays.rs
/*
 * @lc app=leetcode id=4 lang=rust
 *
 * [4] Median of Two Sorted Arrays
 */

// @lc code=start
impl Solution {
    pub fn find_median_sorted_arrays(nums1: Vec<i32>, nums2: Vec<i32>) -> f64 {
        if nums1.len() > nums2.len() {
            Solution::solve(&nums2, &nums1)
        } else {
            Self::solve(&nums1, &nums2)
        }
    }

    pub fn solve(a: &Vec<i32>, b: &Vec<i32>) -> f64 {
        let m = a.len();
        let n = b.len();
        let half = (m + n + 1) / 2;
        let mut min = 0;
        let mut max = m;
        while min <= max {
            let i = (min + max) / 2;
            let j = half - i;
            if i > min && a[i - 1] > b[j] {
                max = i - 1;
            } else if i < max && b[j - 1] > a[i] {
                min = i + 1;
            } else {
                let maxLeft:i32;
                if i == 0 {
                    maxLeft = b[j - 1];
                } else if j == 0 {
                    maxLeft = a[i - 1];
                } else {
                    maxLeft = if a[i - 1] > b [j - 1] { a[i - 1]} else { b[j - 1] };
                }
                if (m + n) % 2 == 1 {
                    return maxLeft as f64;
                }
                let minRight: i32;
                if i == m {
                    minRight = b[j];
                } else if j == n {
                    minRight = a[i];
                } else {
                    minRight = if a[i] < b[j] { a[i] } else { b[j] };
                }

                return (maxLeft + minRight) as f64 / 2.0;
            }
        }

        0.0
    }
}
// @lc code=end

File: 4.median-of-two-sorted-arrays.go
/*
 * @lc app=leetcode id=4 lang=golang
 *
 * [4] Median of Two Sorted Arrays
 */

// @lc code=start
func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	if len(nums1) < len(nums2) {
		return findMedian(nums1, nums2)
	}
	return findMedian(nums2, nums1)
}
func findMedian(nums1 []int, nums2 []int) float64 {
	m := len(nums1)
	n := len(nums2)
	half := (m + n) / 2
	l := 0
	r := m
	for l <= r {
		i := (l + r) / 2
		j := half - i
		if i > 0 && j < n && nums1[i-1] > nums2[j] {
			r = i
			continue
		}
		if i < m && j > 0 && nums1[i] < nums2[j-1] {
			l = i + 1
			continue
		}
		var minRight int
		if i == m {
			minRight = nums2[j]
		} else if j == n {
			minRight = nums1[i]
		} else if nums1[i] < nums2[j] {
			minRight = nums1[i]
		} else {
			minRight = nums2[j]
		}
		if (m+n)%2 == 1 {
			return float64(minRight)
		}
		var maxLeft int
		if i == 0 {
			maxLeft = nums2[j-1]
		} else if j == 0 {
			maxLeft = nums1[i-1]
		} else if nums1[i-1] > nums2[j-1] {
			maxLeft = nums1[i-1]
		} else {
			maxLeft = nums2[j-1]
		}

		return (float64(maxLeft) + float64(minRight)) / 2.0
	}

	return 0.0
}

// @lc code=end

